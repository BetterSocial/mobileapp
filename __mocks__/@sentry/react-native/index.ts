export const addBreadcrumb = jest.fn();
