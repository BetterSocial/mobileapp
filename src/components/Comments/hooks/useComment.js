import React from 'react';
import {iVoteComment, voteComment, voteCommentV2} from '../../../service/vote';

const useComment = ({comment, findCommentAndUpdate, level, updateVote}) => {
  const [totalVote, setTotalVote] = React.useState(
    comment.data?.count_upvote - comment.data?.count_downvote
  );
  const [statusVote, setStatusVote] = React.useState('none');
  const onUpVote = async () => {
    if (statusVote === 'upvote') {
      setTotalVote((prevState) => prevState - 1);
      setStatusVote('none');
    }
    if (statusVote === 'downvote') {
      setTotalVote((prevState) => prevState + 2);
      setStatusVote('upvote');
    }
    if (statusVote === 'none') {
      setTotalVote((prevState) => prevState + 1);
      setStatusVote('upvote');
    }
    const dataVote = {
      reaction_id: comment.id,
      vote: 'upvote'
    };
    onVote(dataVote);
  };
  const onDownVote = async () => {
    if (statusVote === 'upvote') {
      setTotalVote((prevState) => prevState - 2);
      setStatusVote('downvote');
    }
    if (statusVote === 'downvote') {
      setTotalVote((prevState) => prevState + 1);
      setStatusVote('none');
    }
    if (statusVote === 'none') {
      setTotalVote((prevState) => prevState - 1);
      setStatusVote('downvote');
    }
    const dataVote = {
      reaction_id: comment.id,
      vote: 'downvote'
    };
    onVote(dataVote);
  };
  const onVote = async (dataVote) => {
    const result = await voteCommentV2(dataVote);
    if (findCommentAndUpdate) {
      findCommentAndUpdate(comment.id, result.data, level);
    }
    if (updateVote) {
      updateVote(result.data, comment, level);
    }
    iVote();
  };

  const iVote = async () => {
    const result = await iVoteComment(comment.id);
    if (result.code === 200) {
      setStatusVote(result.data.action);
    }
  };

  return {
    totalVote,
    setTotalVote,
    statusVote,
    setStatusVote,
    onVote,
    iVote,
    onUpVote,
    onDownVote
  };
};

export default useComment;
