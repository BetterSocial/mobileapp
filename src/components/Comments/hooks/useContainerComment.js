const useContainerComment = () => {
  const isLast = (index, item, comments) =>
    index === comments.length - 1 && (item.children_counts.comment || 0) === 0;

  const isLastInParent = (index, comments) => index === comments.length - 1;

  const hideLeftConnector = (index, comments) => index === comments.length - 1;

  return {
    isLast,
    isLastInParent,
    hideLeftConnector
  };
};

export default useContainerComment;
