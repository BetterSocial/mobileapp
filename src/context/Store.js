import * as React from 'react';

import {channelReducer, channelState} from './reducers/channelReducer';
import {clientReducer, clientState} from './reducers/clientReducer';
import {discoveryReducer, discoveryState} from './reducers/discoveryReducer';
import {domainReducer, domainState} from './reducers/domainReducer';
import {feedsReducer, feedsState} from './reducers/FeedReducer';
import {followingReducer, followingState} from './reducers/followingReducer';
import {generalComponentReducer, generalComponentState} from './reducers/generalComponentReducer';
import {groupChatReducer, groupChatState} from './reducers/groupChat';
import {initialStartupReducer, initialStartupState} from './reducers/intialStartupReducer';
import {localCommunityReducer, localCommunityState} from './reducers/localComunityReducer';
import {myProfileFeedReducer, myProfileFeedState} from './reducers/myProfileFeedReducer';
import {myProfileReducer, myProfileState} from './reducers/myProfileReducer';
import {newsReducer, newsState} from './reducers/newsReducer';
import {otherProfileFeedReducer, otherProfileFeedState} from './reducers/OtherProfileFeedReducer';
import {topicsReducer, topicsState} from './reducers/topicsReducer';
import {unReadMessageReducer, unReadMessageState} from './reducers/unReadMessageReducer';
import {usersReducer, usersState} from './reducers/userReducer';
import {commentReducer, commentState} from './reducers/commentReducer';

/**
 *
 * @typedef {Object} MainContext
 * @property {import('./reducers/discoveryReducer').DiscoveryState} discovery
 */
const Store = ({children}) => {
  const rootReducer = {
    channel: React.useReducer(channelReducer, channelState),
    client: React.useReducer(clientReducer, clientState),
    discovery: React.useReducer(discoveryReducer, discoveryState),
    feeds: React.useReducer(feedsReducer, feedsState),
    following: React.useReducer(followingReducer, followingState),
    generalComponent: React.useReducer(generalComponentReducer, generalComponentState),
    groupChat: React.useReducer(groupChatReducer, groupChatState),
    localCommunity: React.useReducer(localCommunityReducer, localCommunityState),
    myProfileFeed: React.useReducer(myProfileFeedReducer, myProfileFeedState),
    otherProfileFeed: React.useReducer(otherProfileFeedReducer, otherProfileFeedState),
    news: React.useReducer(newsReducer, newsState),
    profile: React.useReducer(myProfileReducer, myProfileState),
    topics: React.useReducer(topicsReducer, topicsState),
    unReadMessage: React.useReducer(unReadMessageReducer, unReadMessageState),
    users: React.useReducer(usersReducer, usersState),
    domains: React.useReducer(domainReducer, domainState),
    initialStartup: React.useReducer(initialStartupReducer, initialStartupState),
    comments: React.useReducer(commentReducer, commentState)
  };
  return <Context.Provider value={rootReducer}>{children}</Context.Provider>;
};
export const Context = React.createContext(null);
export const useAppContext = () => React.useContext(Context);

export default Store;
