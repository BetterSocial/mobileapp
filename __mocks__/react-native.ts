import * as RN from 'react-native';

RN.Animated.timing = () => ({
    start: () => jest.fn()
})

module.exports = RN