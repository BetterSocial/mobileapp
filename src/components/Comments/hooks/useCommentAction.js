import React from 'react';
import {Alert} from 'react-native';
import {useNavigation, useRoute} from '@react-navigation/core';
import {Context} from '../../../context';
import StringConstant from '../../../utils/string/StringConstant';
import {deleteComment} from '../../../service/comment';

const useCommentAction = () => {
  const [profile] = React.useContext(Context).profile;
  const {params} = useRoute();
  const {goBack} = useNavigation();
  const showAlertDelete = (item, backToPrevPage, callback) => {
    if (profile.myProfile?.user_id === item?.user?.id || item?.is_you) {
      Alert.alert(null, StringConstant.feedDeleteCommentConfirmation, [
        {
          text: 'No, cancel',
          style: 'cancel'
        },
        {
          text: 'Yes',
          style: 'destructive',
          onPress: () => onDeleteCommentClicked(item, backToPrevPage, callback)
        }
      ]);
    }
  };
  const onDeleteCommentClicked = async (item, backToPrevPage, callback) => {
    try {
      const response = await deleteComment(item?.id);

      if (response?.success) {
        if (params.getComment && typeof params.getComment === 'function') {
          params.getComment();
        }
        if (backToPrevPage) {
          goBack();
        }
        if (callback && typeof callback === 'function') {
          console.log('masuk callback');
          callback();
        }
      }
    } catch (e) {
      console.log(e, 'eman');
    }
  };

  return {
    showAlertDelete
  };
};

export default useCommentAction;
