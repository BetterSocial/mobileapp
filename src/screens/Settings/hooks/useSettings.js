import React from 'react';
import Toast from 'react-native-simple-toast';
import {Alert} from 'react-native';
import {useSetRecoilState} from 'recoil';

import StringConstant from '../../../utils/string/StringConstant';
import {Context} from '../../../context';
import {InitialStartupAtom} from '../../../service/initialStartup';
import {clearLocalStorege} from '../../../utils/token';
import {createClient} from '../../../context/actions/createClient';
import {deleteAccount, removeFcmToken} from '../../../service/users';
import {removeAllCache} from '../../../utils/cache';
import {resetProfileFeed} from '../../../context/actions/myProfileFeed';
import {setMainFeeds} from '../../../context/actions/feeds';

const useSettings = () => {
  const [isLoadingDeletingAccount, setIsLoadingDeletingAccount] = React.useState(false);

  const [clientState, dispatch] = React.useContext(Context).client;
  const {client} = clientState;
  const setStartupValue = useSetRecoilState(InitialStartupAtom);

  const [, myProfileDispatch] = React.useContext(Context).myProfileFeed;
  const [, feedDispatch] = React.useContext(Context).feeds;

  const logout = async () => {
    await removeFcmToken();
    removeAllCache();
    resetProfileFeed(myProfileDispatch);
    setMainFeeds([], feedDispatch);
    client?.disconnectUser();
    createClient(null, dispatch);
    clearLocalStorege();
  };

  const handleResponseDelete = async (response) => {
    if (response.status === 'success') {
      logout();
      console.log('');
      Toast.show(StringConstant.profileDeleteAccountSuccess, Toast.SHORT);

      setStartupValue({
        id: null,
        deeplinkProfile: false
      });
    }
    setIsLoadingDeletingAccount(false);
  };

  const doDeleteAccount = async () => {
    setIsLoadingDeletingAccount(true);
    const response = await deleteAccount();
    handleResponseDelete(response);
  };

  const showDeleteAccountAlert = () => {
    Alert.alert(
      StringConstant.profileDeleteAccountAlertTitle,
      StringConstant.profileDeleteAccountAlertMessage,
      [
        {text: StringConstant.profileDeleteAccountAlertCancel, onPress: () => {}, style: 'default'},
        {
          text: StringConstant.profileDeleteAccountAlertSubmit,
          onPress: doDeleteAccount,
          style: 'destructive'
        }
      ]
    );
  };

  return {
    logout,
    isLoadingDeletingAccount,
    setIsLoadingDeletingAccount,
    client,
    setStartupValue,
    myProfileDispatch,
    feedDispatch,
    dispatch,
    doDeleteAccount,
    showDeleteAccountAlert,
    handleResponseDelete
  };
};

export default useSettings;
