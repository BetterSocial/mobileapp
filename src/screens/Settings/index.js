import * as React from 'react';
import VersionNumber from 'react-native-version-number';
import analytics from '@react-native-firebase/analytics';
import {
  Dimensions,
  SafeAreaView,
  StatusBar,
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from 'react-native';
import {useNavigation} from '@react-navigation/core';
import {useSetRecoilState} from 'recoil';

import Header from '../../components/Header';
import Loading from '../Loading';
import ProfileSettingItem from './element/ProfileSettingItem';
import useSettings from './hooks/useSettings';
import {COLORS} from '../../utils/theme';
import {debugAtom} from '../../service/debug';
import {fonts} from '../../utils/fonts';
import {withInteractionsManaged} from '../../components/WithInteractionManaged';

const {width} = Dimensions.get('screen');

const Settings = () => {
  const navigation = useNavigation();
  const [toggleDebug, setToggleDebug] = React.useState(0);
  const setDebugMode = useSetRecoilState(debugAtom);

  const {isLoadingDeletingAccount, logout, showDeleteAccountAlert, setStartupValue} = useSettings();

  React.useEffect(() => {
    analytics().logScreenView({
      screen_class: 'Settings',
      screen_name: 'Settings'
    });
  }, []);

  const goToPage = (pageName) => {
    navigation.navigate(pageName);
  };

  const doLogout = () => {
    logout();
    setStartupValue({
      id: null,
      deeplinkProfile: false
    });
  };

  const turnOnDebugMode = () => {
    if (toggleDebug > 7) {
      setDebugMode(true);
    } else {
      setToggleDebug(toggleDebug + 1);
    }
  };

  return (
    <>
      <StatusBar barStyle="dark-content" translucent={false} />
      <SafeAreaView style={styles.container}>
        <Header title="Settings" isCenter onPress={() => navigation.goBack()} />
        <View style={styles.content}>
          <ProfileSettingItem
            testID="blocked"
            text="Blocked List"
            onPress={() => goToPage('BlockScreen')}
          />
          <ProfileSettingItem
            testID="privacy"
            text="Privacy Policies"
            onPress={() => goToPage('PrivacyPolicies')}
          />
          <ProfileSettingItem
            testID="help"
            text="Help Center"
            onPress={() => goToPage('HelpCenter')}
          />
          <ProfileSettingItem text="Delete Account" onPress={showDeleteAccountAlert} />
          <ProfileSettingItem text="Logout" onPress={doLogout} />
        </View>
        <View>
          <TouchableOpacity onPress={turnOnDebugMode}>
            <View style={styles.versionContainer}>
              <Text style={styles.textVersion}>{`Version ${VersionNumber.appVersion}`}</Text>
            </View>
          </TouchableOpacity>
        </View>

        <Loading visible={isLoadingDeletingAccount} />
      </SafeAreaView>
    </>
  );
};
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.white
  },
  containerHeader: {padding: 16},
  header: {
    flexDirection: 'row',
    justifyContent: 'center',
    marginTop: 6,
    paddingBottom: 16,
    paddingTop: 16,
    position: 'relative'
  },
  textSettings: {
    fontFamily: fonts.inter[600],
    fontWeight: 'bold',
    fontSize: 14,
    color: COLORS.black
  },
  floatLeft: {
    position: 'absolute',
    left: 20,
    top: 20
  },
  content: {
    padding: 20,
    flex: 20
  },
  footer: {
    zIndex: 999
  },
  versionContainer: {
    width,
    height: 40,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center'
  },
  textVersion: {
    fontFamily: fonts.inter[400],
    fontSize: 12,
    color: COLORS.gray400
  }
});
export default withInteractionsManaged(React.memo(Settings));
