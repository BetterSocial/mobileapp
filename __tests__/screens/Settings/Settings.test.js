import React from 'react';
import {act, fireEvent, render} from '@testing-library/react-native';
import Settings from '../../../src/screens/Settings';
import Store from '../../../src/context/Store';
import * as useSetting from '../../../src/screens/Settings/hooks/useSettings';

const mockLogScreenView = jest.fn();
const mockGoBack = jest.fn();
const mockNavigate = jest.fn();
jest.mock('@react-native-firebase/analytics', () => () => ({
  logEvent: jest.fn(),
  logLogin: jest.fn(),
  setUserId: jest.fn(),
  logScreenView: mockLogScreenView
}));

jest.mock('@react-navigation/core', () => ({
  ...jest.requireActual('@react-navigation/core'),
  useNavigation: () => ({goBack: mockGoBack, navigate: mockNavigate}),
  useRoute: () => ({
    params: {}
  })
}));

jest.mock('../../../src/hooks/useAfterInteractions', () => ({
  useAfterInteractions: () => ({
    transitionRef: {current: {animateNextTransition: jest.fn()}},
    interactionsComplete: true
  })
}));

describe('Settings page should run correctly', () => {
  const mockLogout = jest.fn();
  const mockStartupValue = jest.fn();
  const showDeleteAccountAlert = jest.fn();
  beforeEach(() => {
    jest.spyOn(useSetting, 'default').mockImplementation(() => ({
      logout: mockLogout,
      setStartupValue: mockStartupValue,
      showDeleteAccountAlert
    }));
  });

  it('Should match snapshot', () => {
    const {toJSON} = render(<Settings />, {wrapper: Store});
    expect(toJSON).toMatchSnapshot();
  });

  it('back button should run back', () => {
    const {getByTestId} = render(<Settings />, {wrapper: Store});
    act(() => {
      fireEvent.press(getByTestId('backButton'));
    });
    expect(mockGoBack).toHaveBeenCalled();
  });

  it('profileSetting on press should navigate to other page', () => {
    const {getByTestId} = render(<Settings />, {wrapper: Store});
    act(() => {
      fireEvent.press(getByTestId('blocked'));
    });
    expect(mockNavigate).toHaveBeenCalled();
    act(() => {
      fireEvent.press(getByTestId('privacy'));
    });
    expect(mockNavigate).toHaveBeenCalled();
    act(() => {
      fireEvent.press(getByTestId('help'));
    });
    expect(mockNavigate).toHaveBeenCalled();
  });

  it('onLogout should run correctly', () => {
    const {getByTestId} = render(<Settings />, {wrapper: Store});
    act(() => {
      fireEvent.press(getByTestId('logout'));
    });
    expect(mockLogout).toHaveBeenCalled();
    expect(mockStartupValue).toHaveBeenCalled();
  });

  it('delete account should run correctly', () => {
    const {getByTestId} = render(<Settings />, {wrapper: Store});
    act(() => {
      fireEvent.press(getByTestId('delete'));
    });
    expect(showDeleteAccountAlert).toHaveBeenCalled();
  });
});
