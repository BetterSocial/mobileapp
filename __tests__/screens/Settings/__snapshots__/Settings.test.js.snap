// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings page should run correctly Should match snapshot 1`] = `[Function]`;
